File: 2024/day-08/main.cpp
#include <iostream>
#include <map>
#include <set>
#include <utility>
#include <vector>

using namespace std;

void readInput(vector<string> &antennaMap, map<char, set<pair<int, int>>> &antennas)
{
    string line;
    for (size_t row = 0; getline(cin, line); row++) {
        for (size_t col = 0; col < line.size(); col++) {
            char frequency = line[col];
            if (frequency != '.') {
                if (antennas.count(frequency)) {
                    antennas.at(frequency).insert({row, col});
                } else {
                    antennas.insert({frequency, {{row, col}}});
                }
            }
        }
        antennaMap.push_back(line);
    }
}

pair<int, int> getDirectionVector(const pair<int, int> &antenna1, const pair<int, int> &antenna2)
{
    return {antenna2.first - antenna1.first, antenna2.second - antenna1.second};
}

bool isInBounds(const pair<int, int> &location, const vector<string> &map)
{
    return (location.first >= 0 && cmp_less(location.first, map.size()))
           && (location.second >= 0 && cmp_less(location.second, map[0].size()));
}

vector<pair<int, int>> getAntinodes(const pair<int, int> &antenna1,
                                    const pair<int, int> &antenna2,
                                    const vector<string> &antennaMap)
{
    if (antenna1 == antenna2) {
        return {};
    }

    const pair<int, int> directionVector = getDirectionVector(antenna1, antenna2);

    const pair<int, int> antinode1 = {antenna1.first - directionVector.first,
                                      antenna1.second - directionVector.second};
    const pair<int, int> antinode2 = {antenna2.first + directionVector.first,
                                      antenna2.second + directionVector.second};

    vector<pair<int, int>> antinodes;
    if (isInBounds(antinode1, antennaMap)) {
        antinodes.push_back(antinode1);
    }
    if (isInBounds(antinode2, antennaMap)) {
        antinodes.push_back(antinode2);
    }
    return antinodes;
}

vector<pair<int, int>> getResonantHarmonicsAntinodes(const pair<int, int> &antenna1,
                                                     const pair<int, int> &antenna2,
                                                     const vector<string> &antennaMap)
{
    if (antenna1 == antenna2) {
        return {};
    }

    const pair<int, int> directionVector = getDirectionVector(antenna1, antenna2);
    vector<pair<int, int>> antinodes;

    pair<int, int> antinode1 = antenna1;
    do {
        antinodes.push_back(antinode1);
        antinode1.first -= directionVector.first;
        antinode1.second -= directionVector.second;
    } while (isInBounds(antinode1, antennaMap));

    pair<int, int> antinode2 = antenna2;
    do {
        antinodes.push_back(antinode2);
        antinode2.first += directionVector.first;
        antinode2.second += directionVector.second;
    } while (isInBounds(antinode2, antennaMap));

    return antinodes;
}

int main()
{
    vector<string> antennaMap;
    map<char, set<pair<int, int>>> antennas;
    readInput(antennaMap, antennas);

    set<pair<int, int>> uniqueAntinodes;
    set<pair<int, int>> uniqueResonantHarmonicsAntinodes;
    set<pair<pair<int, int>, pair<int, int>>> consideredAntennaCombinations;

    for (const auto &antennasWithSameFrequency : antennas) {
        for (const auto &antenna1 : antennasWithSameFrequency.second) {
            for (const auto &antenna2 : antennasWithSameFrequency.second) {
                const auto orderedPair = make_pair(min(antenna1, antenna2), max(antenna1, antenna2));
                if (consideredAntennaCombinations.contains(orderedPair)) {
                    continue;
                }
                consideredAntennaCombinations.insert(orderedPair);

                const vector<pair<int, int>> foundAntinodes = getAntinodes(antenna1,
                                                                           antenna2,
                                                                           antennaMap);
                for (pair<int, int> antinode : foundAntinodes) {
                    uniqueAntinodes.insert(antinode);
                }
                const vector<pair<int, int>> foundResonantHarmonicsAntinodes
                    = getResonantHarmonicsAntinodes(antenna1, antenna2, antennaMap);
                for (pair<int, int> antinode : foundResonantHarmonicsAntinodes) {
                    uniqueResonantHarmonicsAntinodes.insert(antinode);
                }
            }
        }
    }

    // Part one
    cout << "Number of unique antinodes: " << uniqueAntinodes.size() << endl;

    // Part two
    cout << "Number of unique antinodes when taking resonant harmonics into account: "
         << uniqueResonantHarmonicsAntinodes.size() << endl;
}

File: 2024/go/day-08/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"
)

type position struct {
	row, col int
}

func readInput() (antennas map[rune][]position, dimensions position, err error) {
	content, err := os.ReadFile("input")
	if err != nil {
		return
	}

	lines := strings.Split(strings.TrimSpace(string(content)), "\n")
	antennas = make(map[rune][]position)

	for rowIndex, line := range lines {
		for colIndex, frequency := range line {
			if frequency != '.' {
				antennas[frequency] = append(antennas[frequency], position{row: rowIndex, col: colIndex})
			}
		}
	}

	dimensions = position{row: len(lines), col: len(lines[0])}

	return
}

func isInBounds(antinode, dimensions position) bool {
	return (antinode.row >= 0 && antinode.row < dimensions.row) && (antinode.col >= 0 && antinode.col < dimensions.col)
}

func deltaVector(antenna1, antenna2 position) position {
	return position{
		row: antenna2.row - antenna1.row,
		col: antenna2.col - antenna1.col,
	}
}

func antennaPairAntinodes(antenna1, antenna2, dimensions position) []position {
	deltaVector := deltaVector(antenna1, antenna2)

	antinode1 := position{
		row: antenna1.row - deltaVector.row,
		col: antenna1.col - deltaVector.col,
	}
	antinode2 := position{
		row: antenna2.row + deltaVector.row,
		col: antenna2.col + deltaVector.col,
	}

	var antinodes []position
	if isInBounds(antinode1, dimensions) {
		antinodes = append(antinodes, antinode1)
	}
	if isInBounds(antinode2, dimensions) {
		antinodes = append(antinodes, antinode2)
	}

	return antinodes
}

func antennaPairResonantAntinodes(antenna1, antenna2, dimensions position) []position {
	var resonantAntinodes []position
	deltaVector := deltaVector(antenna1, antenna2)

	// Resonant antinodes
	antinode1 := antenna1
	for isInBounds(antinode1, dimensions) {
		resonantAntinodes = append(resonantAntinodes, antinode1)

		antinode1.row -= deltaVector.row
		antinode1.col -= deltaVector.col
	}

	antinode2 := antenna2
	for isInBounds(antinode2, dimensions) {
		resonantAntinodes = append(resonantAntinodes, antinode2)

		antinode2.row += deltaVector.row
		antinode2.col += deltaVector.col
	}

	return resonantAntinodes
}

func findAllAntinodes(antennas map[rune][]position, dimensions position) (uniqueAntinodes, uniqueResonantAntinodes map[position]struct{}) {
	uniqueAntinodes = make(map[position]struct{})
	uniqueResonantAntinodes = make(map[position]struct{})

	for _, frequencyGroup := range antennas {
		for antenna1Index := 0; antenna1Index < len(frequencyGroup)-1; antenna1Index++ {
			for antenna2Index := antenna1Index + 1; antenna2Index < len(frequencyGroup); antenna2Index++ {
				antenna1 := frequencyGroup[antenna1Index]
				antenna2 := frequencyGroup[antenna2Index]

				foundAntinodes := antennaPairAntinodes(antenna1, antenna2, dimensions)
				foundResonantAntinodes := antennaPairResonantAntinodes(antenna1, antenna2, dimensions)

				for _, antinode := range foundAntinodes {
					uniqueAntinodes[antinode] = struct{}{}
				}

				for _, antinode := range foundResonantAntinodes {
					uniqueResonantAntinodes[antinode] = struct{}{}
				}
			}
		}
	}

	return
}

func main() {
	antennas, dimensions, err := readInput()
	if err != nil {
		log.Fatalln(err)
	}

	uniqueAntinodes, uniqueResonantAntinodes := findAllAntinodes(antennas, dimensions)

	// Part one
	fmt.Println("Number of unique antinodes:", len(uniqueAntinodes))
	// Part two
	fmt.Println("Number of unique antinodes when taking resonant harmonics into account:", len(uniqueResonantAntinodes))
}
